cardano-node/src/Cardano/CLI/Tx/BenchmarkingTxSubmission.hs

```diff
@@ -0,0 +1,353 @@
+          lift . atomically $ putTMVar resp [(a,c) | (a,_,c) <-  send']
+          setRateLimiter env [ c | (_,_,c) <- send']
+          modify (\x -> x { inFlight    = inFlight x Seq.>< send
+                          , notYetSent  = store })
+          noteBusy
+        else do
+          noteIdle
+          lift . atomically $ putTMVar resp []
+          modify (\x -> x { availableOp = Nothing })
+
+        -- deal with acked txs
+        processAcks acked
+
+  checkRateLimiter :: StateT (RWEnv m txid tx) m ()
+  checkRateLimiter = do
+    waitUntil <- gets proceedAfter
+    sleepFor <- (\x -> waitUntil `diffTime` x) <$> lift getMonotonicTime
+    when (sleepFor > 0) . lift $ do
+      traceWith tr $ TraceBenchTxSubRateLimit sleepFor
+      threadDelay sleepFor
+
+  setRateLimiter
+    :: ROEnv txid tx
+    -> [TxSubmit.TxSizeInBytes]
+    -> StateT (RWEnv m txid tx) m ()
+  setRateLimiter env tls = do
+    let txLimit   = (* fromIntegral (length tls)) <$> txNumServiceTime  env
+        sizeLimit = (* fromIntegral (sum    tls)) <$> txSizeServiceTime env
+        limit     = max txLimit sizeLimit
+    flip (maybe (pure ())) limit $ \d -> do
+      waitUntil <- (addTime d) <$> lift getMonotonicTime
+      modify (\x -> x { proceedAfter = waitUntil })
+
+  processAcks :: Word16 -> StateT (RWEnv m txid tx) m ()
+  processAcks acked  =  when (acked > 0) $ do
+    (done, left) <- (Seq.splitAt (fromIntegral acked)) <$> gets inFlight
+    lift . traceWith tr $ TraceBenchTxSubServAck [a | (a,_,_) <- toList done]
+    modify (\x -> x {inFlight = left})
+
+  noteIdle :: StateT (RWEnv m txid tx) m ()
+  noteIdle = do
+--      liftIO $ putStrLn $ "noteIdle"
+    wasBusy <- (== Busy) <$> gets activityState
+    when wasBusy $ do
+      lift . traceWith tr $ TraceBenchTxSubIdle
+--        liftIO $ putStrLn $ "noteIdle, Idle!"
+      modify (\x -> x { activityState = Idle})
+
+  noteBusy :: StateT (RWEnv m txid tx) m ()
+  noteBusy = do -- just can't get those memory write cycles out of my head
+--      liftIO $ putStrLn $ "noteBusy"
+    wasIdle <- (== Idle) <$> gets activityState
+    when wasIdle $ do
+--        liftIO $ putStrLn $ "noteBusy, Busy!"
+      modify (\x -> x { activityState = Busy})
+
+data ActivityState = Idle | Busy deriving (Eq)
+
+-- | The readonly environment
+data ROEnv txid tx = ROEnv
+  { targetBacklog     :: Int -- ^ how many to try to keep in back pocket, >0
+  , txNumServiceTime  :: Maybe DiffTime -- ^ seconds per tx
+  , txSizeServiceTime :: Maybe DiffTime -- ^ seconds per tx octet
+  } deriving Show
+
+defaultROEnv :: ROEnv txid tx
+defaultROEnv = ROEnv
+  { targetBacklog     = 1
+  , txNumServiceTime  = Nothing
+  , txSizeServiceTime = Nothing
+  }
+
+data RWEnv m txid tx = RWEnv
+  { terminating     :: Bool
+  , activityState   :: ActivityState
+  , proceedAfter    :: Time m
+  , availableOp     :: Maybe (Int, LazyTMVar m (Maybe [(txid, TxSubmit.TxSizeInBytes)]))
+  -- ^ the window and the response action
+  , inFlight
```

Comment:
What's going on with this `inFlight` field? Is this valid code?

Reply:
The CI checks are all passing so I'm guessing it's valid, but I'm wondering why there's no type signature?

Reply:
Wow.